sway/input/keyboard.c

```diff
@@ -9,88 +9,92 @@
 #include "sway/commands.h"
 #include "log.h"
 
-static bool keysym_is_modifier(xkb_keysym_t keysym) {
-	switch (keysym) {
-	case XKB_KEY_Shift_L: case XKB_KEY_Shift_R:
-	case XKB_KEY_Control_L: case XKB_KEY_Control_R:
-	case XKB_KEY_Caps_Lock:
-	case XKB_KEY_Shift_Lock:
-	case XKB_KEY_Meta_L: case XKB_KEY_Meta_R:
-	case XKB_KEY_Alt_L: case XKB_KEY_Alt_R:
-	case XKB_KEY_Super_L: case XKB_KEY_Super_R:
-	case XKB_KEY_Hyper_L: case XKB_KEY_Hyper_R:
-		return true;
-	default:
-		return false;
-	}
-}
-
-static size_t pressed_keysyms_length(xkb_keysym_t *pressed_keysyms) {
-	size_t n = 0;
-	for (size_t i = 0; i < SWAY_KEYBOARD_PRESSED_KEYSYMS_CAP; ++i) {
-		if (pressed_keysyms[i] != XKB_KEY_NoSymbol) {
-			++n;
+/**
+ * Update the shortcut model state in response to new input
+ */
+static void update_shortcut_state(struct sway_shortcut_state *state,
+		struct wlr_event_keyboard_key *event, uint32_t new_key,
+		bool last_key_was_a_modifier) {
+	if (event->state == WLR_KEY_PRESSED) {
+		if (last_key_was_a_modifier && state->last_key_index >= 0) {
+			// Last pressed key before this one was a modifier. We nullify
+			// the key id but not the keycode (as that is used for erasure
+			// on release)
+			state->pressed_keys[state->last_key_index] = 0;
+			state->last_key_index = -1;
```

Comment:
This section erases modifier keys once they have been identified as changing the modifier state.
As this is discovered on the keypress after the modifier key, `last_key_index` is stored so that the modifier key can be identified.

As for `last_key_index` is reset: it may be that more than 32 keys are pressed, and then the value
might otherwise persist and cause an incorrect removal from the pressed key list.

Reply:
All right, thanks for the explanation